### PR TITLE
Add Content Switcher component

### LINKED ISSUE #485: Internal Pages Menu
Figma link: https://www.figma.com/file/w60dl1XPUvSaRncv1Utmnb/Openverse-Releases?node-id=25%3A2962

image

## Questions

- Does this need to be a named, standalone component, or just a popover + button in the header component?
- This three dot icon sticks out to me as odd. It's a lighter weight than any of the other icons in the system
- What are the focus + hover styles of these inner links? Related to a similar question in #479

### PR DESCRIPTION
**Pull request by obulat**
## Fixes

Fixes #485 by @zackkrida
This PR adds the Content Switcher to the Header. This is a new version of #591 that does not have merge conflicts. It is also based off of the `filter_sidebar` branch, so it shouldn't be merged before that.

## Description

The `VContentSwitcher` component combines the Internal Page menu and the Content type switcher. 
There is also a `VPageMenu` component used on internal pages (eg. `About`), where only the page menu, and not content switcher, are used.

This component turned out to be very complex.
On mobile screens and on initial SSR, the page menu button does not show, and clicking the content switcher button opens a modal with the Content types menu and the Pages menu.
On larger screens, clicking the Pages menu button (ellipsis icon) opens up a Popover with links to internal pages, and clicking the Content type button opens up a Popover with different content types available. Clicking on the items changes the search content type, and closes the popover.

### Issues
There is a problem with tabbing through the items in the mobile modal. I am not sure why they can't be tabbed through, it worked previously.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [ ] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by krysal — COMMENTED** [review]
This is looking really good! It's noticeable that these components are broad in functionality and styles variants. I want to check some files in more detail to learn more from your approach :) but at a first glance seems pretty complete.

I'm just seeing that the `VPageMenu` doesn't close after clicking on a link, it effectively navigates but the popover/full page modal keeps open, one has to click outside/close the modal to see the target page.

File: src/composables/use-scroll.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      ? throttle(scrollHandler, throttleMs)
+      : scrollHandler
+
+    useEventListener(element, 'scroll', handler, {
+      capture: false,
+      passive: true,
+    })
+  }
+
+  return { y, isScrolled }
```
[the remark this answers is not in the document]
Comment (obulat) [reply src/composables/use-scroll.js]: Their purposes are similar. At first, we were scrolling the whole window, but then we updated the layout so that it consists of a grid, in which the main content and the sidebar can be scrolled separately. This means that we can no longer detect if the header has scrolled or not by using `window`'s scroll offset, and need to get the main content `scrollTop` instead. I did not remove the `use-window-scroll.js` just yet. But if it turns out after all the redesign work that we don't, we should delete `use-window-scroll.js`.

File: src/components/VContentSwitcher/VContentTypes.vue [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import VItemGroup from '~/components/VItemGroup/VItemGroup.vue'
+
+export default {
+  name: 'VContentTypes',
+  components: { VIcon, VItem, VItemGroup },
+  props: {
+    bordered: {
+      type: Boolean,
+      default: true,
+      // !isMinScreenMd
```
Comment (sarayourfriend) [inline src/components/VContentSwitcher/VContentTypes.vue:39]:
```suggestion
```

File: src/components/VContentSwitcher/VMobileMenuModal.vue [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        close()
+      } else {
+        open()
+      }
+    }
+    const selectItem = (item) => {
+      emit('select', item)
+    }
+
+    const isLinkExternal = (item) => !item.link.startsWith('/')
```
Comment (sarayourfriend) [inline src/components/VContentSwitcher/VMobileMenuModal.vue:106]: Nice! This would be something we could add to the `VButton` component as well 😮

File: src/composables/use-content-type.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import allIcon from '~/assets/icons/all-content.svg'
+import audioIcon from '~/assets/icons/audio-content.svg'
+import imageIcon from '~/assets/icons/image-content.svg'
+
+const icons = {
+  all: allIcon,
+  audio: audioIcon,
+  image: imageIcon,
+}
+export default function useContentType() {
```
Comment (sarayourfriend) [inline src/composables/use-content-type.js:14]: I really like this way of encapsulating parts of Vuex into a composable. It feels really clean and creates a much nicer API for the components to work with. As a bonus I suspect this would make it easier to switch to pinia in the future.

File: src/store/search.js [hunk 5]
```diff
@@ -346,7 +346,9 @@ const actions = {
       }
       commit(CLEAR_OTHER_MEDIA_TYPE_FILTERS, { searchType })
     }
-    await dispatch(UPDATE_QUERY_FROM_FILTERS, queryParams)
+    if (queryParams !== {}) {
```
Comment (sarayourfriend) [inline src/store/search.js:349]: This python-ism won't work in JavaScript 😢  If we need to test for an empty object there are two ways we could do it. Either we could test `Object.keys(...).length` and check it for `0` (which I'm assuming has some caveats that I can't think of, it seems like it should work fine) or we could use the `lodash.isEmpty` function.

File: src/utils/search-query-transform.js [hunk 6]
```diff
@@ -216,3 +216,12 @@ export const queryStringToQueryData = (queryString) => {
...
 }
+
+export const areQueriesEqual = (oldQuery, newQuery) => {
+  for (let key of Object.keys(oldQuery)) {
+    if (oldQuery[key] !== newQuery[key]) {
+      return false
+    }
+  }
+  return true
+}
```
Comment (sarayourfriend) [inline src/utils/search-query-transform.js:227]: Maybe this doesn't matter for queries but this wouldn't work if `newQuery` has keys that aren't present in `oldQuery`. For example, this will return `true` for:

```js
areQueriesEqual({}, { foo: 'bar' })
```

We could use `lodash.isEqual` or else do something like is described in [this SO answer](https://stackoverflow.com/a/35951373) after converting both arguments to Map objects or adapting that solution to just work on regular objects and avoid the conversion.

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
Just one blocker on the `areQueriesEqual` function and an invalid object literal comparison, but otherwise tests well and LGTM! Nice work!

File: src/utils/search-query-transform.js (hunk 6 above)
Comment (obulat) [reply src/utils/search-query-transform.js:227]: @sarayourfriend, I added `lodash.isEqual`. I was using this comparison because the queries were supposed to have the same structure, and the same number of properties. In fact, we probably won't need this comparison after the All content page is actually different from the Image page.

**Review by sarayourfriend — APPROVED** [review]
LGTM!

File: src/utils/search-query-transform.js (hunk 6 above)
Comment (sarayourfriend) [reply src/utils/search-query-transform.js:227]: Gotcha! I figured that maybe this specific consideration didn't matter for queries for that reason. Either way works for me!

**Review by krysal — APPROVED** [review]
Works perfectly on a desktop layout! This is an incredible work 🌟 

I noticed some issues on mobile, like the label of the content switcher button not changing, or the Openverse logo text not hiding on non-search pages, but I think those can be safely addressed in separate PR.


https://user-images.githubusercontent.com/9145885/149981543-44829c9a-2cfa-49fe-8b68-5d6fa2949793.mp4
